server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java

```diff
@@ -0,0 +1,184 @@
+
+    builder.build();
+  }
+
+  @Test
+  public void setStatus_throws_NPE_if_status_is_null() {
+    Health.Builder builder = newHealthCheckBuilder();
+
+    expectStatusNotNullNPE();
+
+    builder.setStatus(null);
+  }
+
+  @Test
+  public void getStatus_returns_status_from_builder() {
+    Health underTest = newHealthCheckBuilder().setStatus(randomStatus).build();
+
+    assertThat(underTest.getStatus()).isEqualTo(randomStatus);
+  }
+
+  @Test
+  public void addCause_throws_NPE_if_arg_is_null() {
+    Health.Builder builder = newHealthCheckBuilder();
+
+    expectedException.expect(NullPointerException.class);
+    expectedException.expectMessage("cause can't be null");
+
+    builder.addCause(null);
+  }
+
+  @Test
+  public void addCause_throws_IAE_if_arg_is_empty() {
+    Health.Builder builder = newHealthCheckBuilder();
+
+    expectCauseCannotBeEmptyIAE();
+
+    builder.addCause("");
+  }
+
+  @Test
+  public void addCause_throws_IAE_if_arg_contains_only_spaces() {
+    Health.Builder builder = newHealthCheckBuilder();
+
+    expectCauseCannotBeEmptyIAE();
+
+    builder.addCause(Strings.repeat(" ", 1 + random.nextInt(5)));
+  }
+
+  @Test
+  public void getCause_returns_causes_from_builder() {
+    Health.Builder builder = newHealthCheckBuilder().setStatus(randomStatus);
+    randomCauses.forEach(builder::addCause);
+    Health underTest = builder.build();
+
+    assertThat(underTest.getCauses())
+      .isEqualTo(randomCauses);
+  }
+
+  @Test
+  public void green_constant() {
+    assertThat(Health.GREEN).isEqualTo(newHealthCheckBuilder().setStatus(Health.Status.GREEN).build());
+  }
+
+  @Test
+  public void equals_is_based_on_status_and_causes() {
+    Health.Builder builder1 = newHealthCheckBuilder();
+    Health.Builder builder2 = newHealthCheckBuilder();
+
+    builder1.setStatus(randomStatus);
+    builder2.setStatus(randomStatus);
+    randomCauses.forEach(s -> {
+      builder1.addCause(s);
+      builder2.addCause(s);
+    });
+
+    assertThat(builder1.build())
+      .isEqualTo(builder1.build())
+      .isEqualTo(builder2.build())
+      .isEqualTo(builder2.build());
+  }
+
+  @Test
+  public void not_equals_to_null_nor_other_type() {
+    assertThat(Health.GREEN).isNotEqualTo(null);
+    assertThat(Health.GREEN).isNotEqualTo(new Object());
+    assertThat(Health.GREEN).isNotEqualTo(Health.Status.GREEN);
+  }
+
+  @Test
+  public void hashcode_is_based_on_status_and_causes() {
+    Health.Builder builder1 = newHealthCheckBuilder();
+    Health.Builder builder2 = newHealthCheckBuilder();
+    builder1.setStatus(randomStatus);
+    builder2.setStatus(randomStatus);
+    randomCauses.forEach(s -> {
+      builder1.addCause(s);
+      builder2.addCause(s);
+    });
+
+    assertThat(builder1.build().hashCode())
+      .isEqualTo(builder1.build().hashCode())
+      .isEqualTo(builder2.build().hashCode())
+      .isEqualTo(builder2.build().hashCode());
+  }
+
+  @Test
+  public void verify_toString() {
```

Comment:
This test is over-complex regarding the value and the code being tested. It's just testing a trivial toString() method. Testing a single state is far enough, for example GREEN and 2 messages. No need to test all the cases (0, 1 or more causes).

Reply:
problem with 2 messages is that causes are stored in a set and order in a set is not predictive (even in the same JVM, I got the error, it surprised me)